File: backend/src/analyzer/file_filters.py
import re
from pathlib import Path

# Constants for file filtering
AI_BATCH_LOGIC_EXCLUDED_EXTS = {
    ".md", ".markdown", ".txt", ".rst", ".log",
    ".json", ".yml", ".yaml", ".toml", ".ini", ".cfg", ".conf",
    ".css", ".scss", ".sass", ".less", ".xml", ".svg", ".lock",
}

AI_BATCH_LOGIC_PREFERRED_EXTS = {
    ".py", ".ts", ".tsx", ".js", ".jsx", ".java", ".go", ".rs",
    ".cs", ".cpp", ".c", ".h", ".hpp", ".kt", ".swift", ".rb", ".php",
}

AI_BATCH_LOGIC_EXCLUDED_BASENAMES = {
    "package.json", "package-lock.json", "pnpm-lock.yaml", "yarn.lock",
    "tsconfig.json", "jsconfig.json", "next.config.mjs", "next.config.js",
    "vite.config.ts", "vite.config.js", "docker-compose.yml", "docker-compose.yaml",
    "dockerfile", "readme.md",
}

AI_BATCH_LOGIC_EXCLUDED_PATH_MARKERS = (
    "/docs/", "/doc/", "/assets/", "/styles/", "/css/", "/migrations/",
    "/scripts/", "/config/", "/settings/", "/.github/",
)

AI_BATCH_TEST_PATH_RE = re.compile(
    r"(?:^|[\\/])(?:test|tests|__tests__|spec)(?:[\\/]|$)|(?:^|[._-])(test|spec)(?:[._-]|$)",
    re.IGNORECASE,
)

AI_BATCH_GENERATED_OR_PACKAGE_PATH_MARKERS = {
    ".next",
    ".electron",
    ".cache",
    ".pnpm",
    ".yarn",
    ".turbo",
    ".parcel-cache",
    ".svelte-kit",
    ".nuxt",
    ".output",
    ".vercel",
    "node_modules",
    "site-packages",
    "dist-packages",
    "vendor",
}


def is_test_path(path_value: str) -> bool:
    return bool(AI_BATCH_TEST_PATH_RE.search(path_value or ""))


def is_generated_or_package_path(path_value: str) -> bool:
    normalized = str(path_value or "").replace("\\", "/").strip().lower().strip("/")
    if not normalized:
        return False

    parts = [segment for segment in normalized.split("/") if segment]
    if not parts:
        return False

    if any(segment in AI_BATCH_GENERATED_OR_PACKAGE_PATH_MARKERS for segment in parts):
        return True

    for segment in parts[:-1]:
        if segment.startswith(".") and len(segment) > 1:
            return True

    return False


def is_non_implementation_path(path_value: str) -> bool:
    normalized = "/" + str(path_value or "").replace("\\", "/").lower().lstrip("/")
    basename = Path(normalized).name
    ext = Path(basename).suffix.lower()

    if basename in AI_BATCH_LOGIC_EXCLUDED_BASENAMES:
        return True
    if ext in AI_BATCH_LOGIC_EXCLUDED_EXTS:
        return True
    if any(marker in normalized for marker in AI_BATCH_LOGIC_EXCLUDED_PATH_MARKERS):
        return True
    if basename.endswith(".config.ts") or basename.endswith(".config.js"):
        return True
    if basename.endswith(".config.mjs") or basename.endswith(".config.cjs"):
        return True
    return False

File: frontend/lib/file-filters.ts
/**
 * Shared utility for determining whether a file path is a test file,
 * generated package file, or non-implementation file.
 */

const TEST_PATH_RE = /(^|[\\/])(test|tests|__tests__|spec)([\\/]|$)|(^|[._-])(test|spec)([._-]|$)/i;

const GENERATED_OR_PACKAGE_MARKERS = new Set([
  ".next",
  ".electron",
  ".cache",
  ".pnpm",
  ".yarn",
  ".turbo",
  ".parcel-cache",
  ".svelte-kit",
  ".nuxt",
  ".output",
  ".vercel",
  "node_modules",
  "vendor",
  "site-packages",
  "dist-packages",
]);

const NON_IMPLEMENTATION_BASENAMES = new Set([
  "package.json",
  "package-lock.json",
  "pnpm-lock.yaml",
  "yarn.lock",
  "tsconfig.json",
  "jsconfig.json",
  "next.config.mjs",
  "next.config.js",
  "vite.config.ts",
  "vite.config.js",
  "docker-compose.yml",
  "docker-compose.yaml",
  "dockerfile",
  "readme.md",
]);

const NON_IMPLEMENTATION_EXTS = new Set([
  ".md", ".markdown", ".txt", ".rst", ".log",
  ".json", ".yml", ".yaml", ".toml", ".ini", ".cfg", ".conf",
  ".css", ".scss", ".sass", ".less", ".xml", ".svg", ".lock",
]);

const NON_IMPLEMENTATION_PATH_MARKERS = [
  "/docs/",
  "/doc/",
  "/assets/",
  "/styles/",
  "/css/",
  "/migrations/",
  "/scripts/",
  "/config/",
  "/settings/",
  "/.github/",
];

/**
 * Returns true if the path is likely a test or spec file.
 */
export const isTestPath = (pathValue?: string | null): boolean => {
  if (!pathValue) return false;
  return TEST_PATH_RE.test(pathValue);
};

/**
 * Returns true if the path is part of a generated directory, build output, or dependency package.
 */
export const isGeneratedOrPackagePath = (pathValue?: string | null): boolean => {
  if (!pathValue) return false;
  const normalized = pathValue.replace(/\\/g, "/").toLowerCase().replace(/^\/+|\/+$/g, "");
  if (!normalized) return false;

  const parts = normalized.split("/").filter(Boolean);
  
  if (parts.some((segment) => GENERATED_OR_PACKAGE_MARKERS.has(segment))) {
    return true;
  }

  // Check if any parent directory is a hidden folder
  return parts.slice(0, -1).some((segment) => segment.startsWith(".") && segment.length > 1);
};

/**
 * Returns true if the path is a config, doc, asset or other non-implementation file.
 */
export const isNonImplementationPath = (pathValue?: string | null): boolean => {
  if (!pathValue) return false;
  const normalized = pathValue.replace(/\\/g, "/").toLowerCase().replace(/^\/+|\/+$/g, "");
  if (!normalized) return false;

  const parts = normalized.split("/").filter(Boolean);
  const basename = parts[parts.length - 1] ?? "";

  if (NON_IMPLEMENTATION_BASENAMES.has(basename)) return true;
  
  if (
    basename.endsWith(".config.ts") || 
    basename.endsWith(".config.js") || 
    basename.endsWith(".config.mjs") || 
    basename.endsWith(".config.cjs")
  ) {
    return true;
  }
  
  const normalizedForMarker = `/${normalized}`;
  if (NON_IMPLEMENTATION_PATH_MARKERS.some((marker) => normalizedForMarker.includes(marker))) {
    return true;
  }

  const dotIdx = basename.lastIndexOf(".");
  const ext = dotIdx >= 0 ? basename.slice(dotIdx) : "";
  return NON_IMPLEMENTATION_EXTS.has(ext);
};
